File: projects/group/streamlit_app/guardian/data/scam_db.py
"""Scam signals database — 1:1 port of ``app/lib/data/scam_db.dart``.

Loads a CSV of (type, value, weight, tag, note) rows and exposes filtered
iterators for blocklisted numbers, phishing domains, and scam keywords.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Iterable


class ScamEntryType(str, Enum):
    NUMBER = "number"
    DOMAIN = "domain"
    KEYWORD = "keyword"


@dataclass(frozen=True)
class ScamEntry:
    type: ScamEntryType
    value: str
    weight: float
    tag: str
    note: str


class ScamDatabase:
    def __init__(self, entries: list[ScamEntry]) -> None:
        self.entries = entries

    @classmethod
    def from_csv(cls, raw: str) -> "ScamDatabase":
        """Parse the CSV. First row is the header; empty rows skipped."""
        return cls(_parse_csv_entries(raw))

    @classmethod
    def from_csvs(cls, base_raw: str, runtime_raw: str | None = None) -> "ScamDatabase":
        """Parse and merge base + runtime CSV rows.

        Runtime entries are appended and deduplicated by ``(type, value)``.
        """
        merged: list[ScamEntry] = []
        seen: set[tuple[ScamEntryType, str]] = set()
        for entry in [*_parse_csv_entries(base_raw), *_parse_csv_entries(runtime_raw or "")]:
            key = (entry.type, entry.value)
            if key in seen:
                continue
            seen.add(key)
            merged.append(entry)
        return cls(merged)

    def bad_numbers(self) -> Iterable[ScamEntry]:
        return (e for e in self.entries if e.type is ScamEntryType.NUMBER)

    def bad_domains(self) -> Iterable[ScamEntry]:
        return (e for e in self.entries if e.type is ScamEntryType.DOMAIN)

    def keywords(self) -> Iterable[ScamEntry]:
        return (e for e in self.entries if e.type is ScamEntryType.KEYWORD)


def _parse_csv_entries(raw: str) -> list[ScamEntry]:
    lines = [line for line in raw.splitlines() if line.strip()]
    if not lines:
        return []

    out: list[ScamEntry] = []
    # Skip header.
    for line in lines[1:]:
        parts = line.split(",")
        if len(parts) < 4:
            continue
        raw_type = parts[0].strip()
        try:
            entry_type = ScamEntryType(raw_type)
        except ValueError:
            continue
        try:
            weight = float(parts[2].strip())
        except ValueError:
            weight = 0.5
        out.append(
            ScamEntry(
                type=entry_type,
                value=parts[1].strip().lower(),
                weight=weight,
                tag=parts[3].strip(),
                note=",".join(parts[4:]).strip() if len(parts) > 4 else "",
            )
        )
    return out

File: projects/group/legacy-flutter/app/lib/data/scam_db.dart
import 'package:riverpod/riverpod.dart';
import 'package:meta/meta.dart';

enum ScamEntryType { number, domain, keyword }

@immutable
class ScamEntry {
  const ScamEntry({
    required this.type,
    required this.value,
    required this.weight,
    required this.tag,
    required this.note,
  });

  final ScamEntryType type;
  final String value;
  final double weight;
  final String tag;
  final String note;
}

class ScamDatabase {
  ScamDatabase(this.entries);

  final List<ScamEntry> entries;

  static ScamDatabase fromCsv(String raw) {
    final lines = raw.split('\n').where((l) => l.trim().isNotEmpty).toList();
    final out = <ScamEntry>[];
    for (var i = 1; i < lines.length; i++) {
      final parts = lines[i].split(',');
      if (parts.length < 4) continue;
      final type = switch (parts[0].trim()) {
        'number' => ScamEntryType.number,
        'domain' => ScamEntryType.domain,
        'keyword' => ScamEntryType.keyword,
        _ => null,
      };
      if (type == null) continue;
      out.add(ScamEntry(
        type: type,
        value: parts[1].trim().toLowerCase(),
        weight: double.tryParse(parts[2].trim()) ?? 0.5,
        tag: parts[3].trim(),
        note: parts.length > 4 ? parts.sublist(4).join(',').trim() : '',
      ));
    }
    return ScamDatabase(out);
  }

  Iterable<ScamEntry> badNumbers() =>
      entries.where((e) => e.type == ScamEntryType.number);
  Iterable<ScamEntry> badDomains() =>
      entries.where((e) => e.type == ScamEntryType.domain);
  Iterable<ScamEntry> keywords() =>
      entries.where((e) => e.type == ScamEntryType.keyword);
}

final scamDatabaseProvider = Provider<ScamDatabase>((ref) {
  throw UnimplementedError(
    'scamDatabaseProvider must be overridden in ProviderScope',
  );
});
